src/main/java/com/hivemq/persistence/local/memory/ClientSessionMemoryLocalPersistence.java

```diff
@@ -0,0 +1,531 @@
+            currentMemorySize.addAndGet(newEntry.getEstimatedSize());
+
+            return newEntry;
+        });
+    }
+
+    @Override
+    public @NotNull ClientSession disconnect(
+            final @NotNull String clientId,
+            final long timestamp,
+            final boolean sendWill,
+            final int bucketIndex,
+            final long expiry) {
+
+        ThreadPreConditions.startsWith(SINGLE_WRITER_THREAD_PREFIX);
+
+        final Map<String, PersistenceEntry<ClientSession>> bucket = getBucket(bucketIndex);
+
+        final ClientSession storedSession = bucket.compute(clientId, (ignored, oldEntry) -> {
+
+            if (oldEntry == null) {
+                // we create a tombstone here
+                final ClientSession clientSession = new ClientSession(false, configuredSessionExpiryInterval);
+                final PersistenceEntry<ClientSession> persistenceEntry = new PersistenceEntry<>(clientSession, timestamp);
+                currentMemorySize.addAndGet(persistenceEntry.getEstimatedSize());
+                return persistenceEntry;
+            }
+
+            currentMemorySize.addAndGet(-oldEntry.getEstimatedSize());
+            final ClientSession oldSession = oldEntry.getObject();
+            final ClientSession newSession;
+            if (sendWill) {
+                newSession = oldSession;
+            } else {
+                removeWillReference(oldSession);
+                newSession = oldSession.copyWithoutWill();
+            }
+
+            if (expiry != SESSION_EXPIRY_NOT_SET) {
+                newSession.setSessionExpiryInterval(expiry);
+            }
+
+            if (newSession.isConnected() && !persistent(newSession)) {
+                sessionsCount.decrementAndGet();
+            }
+
+            newSession.setConnected(false);
+            checkForPayloadExistence(newSession, false);
+
+            final PersistenceEntry<ClientSession> newEntry = new PersistenceEntry<>(newSession, timestamp);
+            currentMemorySize.addAndGet(newEntry.getEstimatedSize());
+            return newEntry;
+        }).getObject().deepCopyWithoutPayload();
+
+
+        dereferenceWillPayload(storedSession);
+        return storedSession;
+    }
+
+    @Override
+    public @NotNull Set<String> getAllClients(final int bucketIndex) {
+
+        final Map<String, PersistenceEntry<ClientSession>> bucket = getBucket(bucketIndex);
+
+        return ImmutableSet.copyOf(bucket.keySet());
+    }
+
+    @ExecuteInSingleWriter
+    @Override
+    public void removeWithTimestamp(final @NotNull String clientId, final int bucketIndex) {
+
+        final Map<String, PersistenceEntry<ClientSession>> bucket = getBucket(bucketIndex);
+        final PersistenceEntry<ClientSession> remove = bucket.remove(clientId);
+        if (remove != null) {
+            final ClientSession clientSession = remove.getObject();
+            if (persistent(clientSession) || clientSession.isConnected()) {
```

Comment:
this is use multiple times, please add a helper method, or example called hasSession

Reply:
not used that much after my changes.